lib/puppet/defaults.rb

```diff
@@ -1074,9 +1109,16 @@ def self.initialize_default_settings!(settings)
       :desc    => "The name to use the Certificate Authority certificate.",
     },
     :cadir => {
-      :default => "$ssldir/ca",
+      :default => lambda { default_cadir },
       :type => :directory,
       :desc => "The root directory for the certificate authority.",
+      :call_hook => :on_initialize_and_write,
+      :hook => proc do |value|
+        if value.start_with?(Puppet[:ssldir])
+          Puppet.log_ca_migration_warning
```

Comment:
nit: can we call `log_ca_migration_warning` without the `Puppet`. It's kinda weird that Puppet is where that method gets defined and the existing methods that I saw didn't do this.

Reply:
We can't call it when it is a `hook` and a proc, but it does work when it is a lambda for the default. I suppose we could just use a lambda instead for the hook...

Reply:
oh, that's fine